### PR TITLE
SuiNode/Authority state are now aware of what protocol versions they support.

### PR DESCRIPTION
**Pull request by mystenmark**
Also, some basic testing infrastructure, and a simple panic test for when the validator encounters an unsupported protocol.

### CODE REVIEW

File: crates/sui-core/src/authority.rs [hunk 1]
```diff
@@ -1578,6 +1598,7 @@ impl AuthorityState {
         let state = AuthorityState::new(
```
Comment (amnn) [inline crates/sui-core/src/authority.rs:1598]: Slightly unrelated to this PR, but is the comment above accurate anymore?

File: crates/sui-protocol-config/src/lib.rs [hunk 2]
```diff
@@ -32,6 +38,46 @@ impl ProtocolVersion {
...
+    }
+}
+
+/// Models the set of protocol versions supported by a validator.
+/// The `sui-node` binary will always use the SYSTEM_DEFAULT constant, but for testing we need
+/// to be able to inject arbitrary versions into SuiNode.
+#[derive(Debug, Clone, Copy)]
+pub struct SupportedProtocolVersions {
+    min: ProtocolVersion,
+    max: ProtocolVersion,
```
Comment (amnn) [inline crates/sui-protocol-config/src/lib.rs:61]: `std::ops::Range<ProtocolVersion>`?

**Review by amnn — APPROVED** [review]
The way supported protocol versions are configured was a little confusing to follow for me, i.e. that the version for the actual node was configurable at runtime, but the version in tests is a constant that gets tweaked by a feature flag.  I would have expected it to be the other way around (the supported protocol version for the node binary was fairly fixed, and the test one was configurable at runtime) was there a reason that this wasn't possible?

**Comment by mystenmark** [comment]
> The way supported protocol versions are configured was a little confusing to follow for me, i.e. that the version for the actual node was configurable at runtime, but the version in tests is a constant that gets tweaked by a feature flag. I would have expected it to be the other way around (the supported protocol version for the node binary was fairly fixed, and the test one was configurable at runtime) was there a reason that this wasn't possible?

I agree its a bit odd. Here's my reasoning:
- In order to test upgrades, we not only need to configure the versions supported by the node, we also need for another version to actually exist. So, if don't tweak the MAX version with a compile flag, we would actually have to ship a dummy version in the released binary.
- I want to be able to test heterogenous protocol support (not all validators support the same version), so that requires that the versions supported by a node are configured at runtime, not compile time.
- Of course, we'll never configure the version support dynamically in a released binary, hence the code in sui-node/src/main.rs

File: crates/sui-core/src/authority.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-core/src/authority.rs:1598]: definitely not!

File: crates/sui-protocol-config/src/lib.rs (hunk 2 above)
Comment (mystenmark) [reply crates/sui-protocol-config/src/lib.rs:61]: Unfortunately Range/RangeInclusive are not Copy, so I would have to litter all code with `.clone()` just to move a pair a `u64`s around.

**Pull request merged**